tests/fetchGitSubmodules.sh

```diff
@@ -0,0 +1,63 @@
+source common.sh
+
+set -u
+
+if [[ -z $(type -p git) ]]; then
+    echo "Git not installed; skipping Git submodule tests"
+    exit 99
+fi
+
+clearStore
+
+rootRepo=$TEST_ROOT/gitSubmodulesRoot
+subRepo=$TEST_ROOT/gitSubmodulesSub
+
+rm -rf ${rootRepo} ${subRepo} $TEST_HOME/.cache/nix/gitv2
+
+initGitRepo() {
+    git init $1
+    git -C $1 config user.email "foobar@example.com"
+    git -C $1 config user.name "Foobar"
+}
+
+addGitContent() {
+    echo "lorem ipsum" > $1/content
+    git -C $1 add content
+    git -C $1 commit -m "Initial commit"
+}
+
+initGitRepo $subRepo
+addGitContent $subRepo
+
+initGitRepo $rootRepo
+
+git -C $rootRepo submodule init
+git -C $rootRepo submodule add $subRepo sub
+git -C $rootRepo add sub
+git -C $rootRepo commit -m "Add submodule"
+
+rev=$(git -C $rootRepo rev-parse HEAD)
+
+pathWithoutSubmodules=$(nix eval --raw "(builtins.fetchGit { url = file://$rootRepo; rev = \"$rev\"; }).outPath")
+pathWithSubmodules=$(nix eval --raw "(builtins.fetchGit { url = file://$rootRepo; rev = \"$rev\"; submodules = true; }).outPath")
```

Comment:
The test blows up if passing `ref = "refs/heads/master"` to fetchGit with `submodules = true`:

```
++ nix eval --raw '(builtins.fetchGit { url = file:///run/user/1000/nix-test/gitSubmodulesRoot; ref = "refs/heads/master"; rev = "96b77ed66ad52090eebe12ff631e79d3c30d6ba1"; submodules = true; }).outPath'
fatal: couldn't find remote ref refs/heads/master
fatal: the remote end hung up unexpectedly
error: program 'git' failed with exit code 128
```

If using `url = /path/to/repo` (instead of `url = file:///path/to/repo`) the error becomes what I see in real life:

```
++ nix eval --raw '(builtins.fetchGit { url = /run/user/1000/nix-test/gitSubmodulesRoot; ref = "refs/heads/master"; rev = "90ccb6a7d2a59893f0c7283dee98396d3149ea62"; submodules = true; }).outPath'
fatal: Refusing to fetch into current branch refs/heads/master of non-bare repository
error: program 'git' failed with exit code 128
fatal: the remote end hung up unexpectedly
``` 

I wonder why `file://...` makes a difference. But anyway, I experience the error also with git ssh url, so it has nothing to do with local file access.

Reply:
I think this fixes it:

```patch
--- a/src/libexpr/primops/fetchGit.cc
+++ b/src/libexpr/primops/fetchGit.cc
@@ -179,7 +179,7 @@ GitInfo exportGit(ref<Store> store, const std::string & uri,
 
         runProgram("git", true, { "init", tmpDir, "--separate-git-dir", tmpGitDir });
         runProgram("git", true, { "-C", tmpDir, "fetch", "--quiet", "--force",
-                                  "--", cacheDir, fmt("%s:%s", *ref, *ref) });
+                                  "--", cacheDir, fmt("%s", *ref) });
 
         runProgram("git", true, { "-C", tmpDir, "checkout", "--quiet", "FETCH_HEAD" });
         runProgram("git", true, { "-C", tmpDir, "remote", "add", "origin", uri });
```